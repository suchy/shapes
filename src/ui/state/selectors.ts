import { State } from './state';

export const filtersSelector = (state: State) => ({
  colorFilters: state.colorFilters,
  shapeFilters: state.shapeFilters
});

export const shapesSelector = (state: State) => {
  const { colorFilters, shapeFilters } = state;

  const shapes = state.shapes.filter(
    (shape) =>
      colorFilters.includes(shape.color) && shapeFilters.includes(shape.shape)
  );

  return shapes;
};

export const loadingSelector = (state: State) => ({
  isFetching: state.isFetching,
  isFetchingError: state.isFetchingError,
  isFetchingDone: state.isFetchingDone
});
