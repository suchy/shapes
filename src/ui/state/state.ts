import { createAction, createReducer, configureStore } from '@reduxjs/toolkit';

import { Shape, ShapeColors, ShapeShapes } from '../../domain/shapes/shape';
import { getShapes as fetchShapes } from '../../domain/shapes/use-cases/get-shapes';

export const getShapes = () => async (dispatch: any) => {
  try {
    dispatch(getShapesInit());
    const shapes = await fetchShapes();
    dispatch(getShapesSuccess(shapes));
  } catch {
    dispatch(getShapesFail());
  }
};

export const getShapesInit = createAction('GET_SHAPES_INIT');

export const getShapesSuccess = createAction<Shape[]>('GET_SHAPES_SUCCESS');

export const getShapesFail = createAction('GET_SHAPES_FAIL');

export const toggleColorFilter = createAction<ShapeColors>(
  'TOGGLE_COLOR_FILTER'
);

export const toggleShapeFilter = createAction<ShapeShapes>(
  'TOGGLE_SHAPE_FILTER'
);

export interface State {
  isFetching: boolean;
  isFetchingError: boolean;
  isFetchingDone: boolean;
  shapes: Shape[];
  colorFilters: string[];
  shapeFilters: string[];
}

const initialState: State = {
  isFetching: false,
  isFetchingError: false,
  isFetchingDone: false,
  colorFilters: Object.keys(ShapeColors),
  shapeFilters: Object.keys(ShapeShapes),
  shapes: []
};

const reducer = createReducer(initialState, (builder) => {
  builder
    .addCase(getShapesInit, (state) => ({
      ...state,
      isFetching: true,
      isFetchingError: false,
      isFetchingDone: false
    }))

    .addCase(getShapesFail, (state) => ({
      ...state,
      isFetching: false,
      isFetchingError: true,
      isFetchingDone: true
    }))

    .addCase(getShapesSuccess, (state, { payload: shapes }) => ({
      ...state,
      isFetching: false,
      isFetchingError: false,
      isFetchingDone: true,
      shapes
    }))

    .addCase(toggleColorFilter, (state, { payload: filter }) => {
      let colorFilters = [...state.colorFilters];

      const isFilterEnabled = colorFilters.includes(filter);

      if (isFilterEnabled) {
        colorFilters = colorFilters.filter((f) => f !== filter);
      } else {
        colorFilters.push(filter);
      }

      if (!colorFilters.length) {
        colorFilters = Object.keys(ShapeColors);
      }

      const newState = { ...state, colorFilters };
      return newState;
    })

    .addCase(toggleShapeFilter, (state, { payload: filter }) => {
      let shapeFilters = [...state.shapeFilters];

      const isFilterEnabled = shapeFilters.includes(filter);

      if (isFilterEnabled) {
        shapeFilters = shapeFilters.filter((f) => f !== filter);
      } else {
        shapeFilters.push(filter);
      }

      if (!shapeFilters.length) {
        shapeFilters = Object.keys(ShapeShapes);
      }

      const newState = { ...state, shapeFilters };
      return newState;
    });
});

export const store = configureStore({ reducer });
