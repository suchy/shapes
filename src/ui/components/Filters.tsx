import { MouseEvent } from 'react';
import { useDispatch } from 'react-redux';
import styled from 'styled-components';

import { useShallowSelector } from '../hooks/use-shallow-selector';

import { toggleColorFilter, toggleShapeFilter } from '../state/state';
import { filtersSelector } from '../state/selectors';

import { ShapeColors, ShapeShapes } from '../../domain/shapes/shape';

const FiltersContainer = styled.div`
  @media (min-width: 1024px) {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-bottom: 8rem;
  }
`;

const FilltersList = styled.ul`
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  margin-bottom: 2.2rem;

  @media (min-width: 1024px) {
    justify-content: flex-start;
    margin-bottom: 0;
  }
`;

const Filter = styled.li<{ isSelected: boolean }>`
  cursor: pointer;
  margin-bottom: 1.8rem;

  @media (min-width: 1024px) {
    margin-bottom: 0;
  }
`;

const ShapeFilter = styled(Filter)`
  margin-right: 2.1rem;
  font-size: 2.1rem;
  font-style: italic;
  font-weight: bold;
  text-transform: uppercase;
  ${({ isSelected }) => !isSelected && 'color: #828282'};

  &:last-child {
    margin-right: 0;
  }
`;

const ColorFilter = styled(Filter)<{ color: ShapeColors }>`
  width: 3.2rem;
  height: 3.2rem;
  border: solid 0.2rem #000;
  margin-left: 1.4rem;
  margin-right: 1.4rem;
  background: ${({ color }) => color};
  overflow: hidden;
  text-indent: 100%;
  ${({ color, isSelected }) => !isSelected && `border-color: ${color}`};

  &:first-child {
    margin-left: 0;
  }

  &:last-child {
    margin-right: 0;
  }
`;

export const Filters = () => {
  const dispatch = useDispatch();
  const { colorFilters, shapeFilters } = useShallowSelector(filtersSelector);

  const colors = Object.values(ShapeColors);
  const shapes = Object.values(ShapeShapes);

  const isSelected = (filter: string) =>
    colorFilters.includes(filter) || shapeFilters.includes(filter);

  const handleShapeFilterClick = (filter: ShapeShapes) => (
    event: MouseEvent<HTMLLIElement>
  ) => {
    event.preventDefault();
    dispatch(toggleShapeFilter(filter));
  };

  const handleColorFilterClick = (filter: ShapeColors) => (
    event: MouseEvent<HTMLLIElement>
  ) => {
    event.preventDefault();
    dispatch(toggleColorFilter(filter));
  };

  return (
    <FiltersContainer>
      <FilltersList>
        {shapes.map((shape) => (
          <ShapeFilter
            key={shape}
            isSelected={isSelected(shape)}
            title={shape}
            onClick={handleShapeFilterClick(shape)}
            data-testid={`ShapeFilter-${shape}`}
          >
            {shape}
          </ShapeFilter>
        ))}
      </FilltersList>

      <FilltersList>
        {colors.map((color) => (
          <ColorFilter
            key={color}
            color={color}
            isSelected={isSelected(color)}
            title={color}
            onClick={handleColorFilterClick(color)}
            data-testid={`ColorFilter-${color}`}
          >
            {color}
          </ColorFilter>
        ))}
      </FilltersList>
    </FiltersContainer>
  );
};
