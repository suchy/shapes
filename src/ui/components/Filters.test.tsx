import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';

import { store } from '../state/state';

import { Filters } from './Filters';

describe('Filters', () => {
  it('should match snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <Filters />
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should toggle color selection', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Filters />
      </Provider>
    );

    const redFilter = getByTestId('ColorFilter-red');
    fireEvent.click(redFilter);

    setTimeout(() => {
      expect(redFilter).toHaveStyle('border: solid 0.2rem red');
    });
  });

  it('should toggle shape selection', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Filters />
      </Provider>
    );

    const squareFilter = getByTestId('ShapeFilter-square');
    fireEvent.click(squareFilter);

    setTimeout(() => {
      expect(squareFilter).toHaveStyle('color: #828282');
    });
  });
});
