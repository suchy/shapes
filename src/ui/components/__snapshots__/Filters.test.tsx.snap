// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters should match snapshot 1`] = `
<div
  class="sc-bdfBwQ bKpwki"
>
  <ul
    class="sc-gsTCUz nUtTt"
  >
    <li
      class="sc-dlfnbm sc-hKgILt jdwhnD kCTcbZ"
      data-testid="ShapeFilter-round"
      title="round"
    >
      round
    </li>
    <li
      class="sc-dlfnbm sc-hKgILt jdwhnD kCTcbZ"
      data-testid="ShapeFilter-square"
      title="square"
    >
      square
    </li>
    <li
      class="sc-dlfnbm sc-hKgILt jdwhnD kCTcbZ"
      data-testid="ShapeFilter-triangle"
      title="triangle"
    >
      triangle
    </li>
    <li
      class="sc-dlfnbm sc-hKgILt jdwhnD kCTcbZ"
      data-testid="ShapeFilter-oval"
      title="oval"
    >
      oval
    </li>
    <li
      class="sc-dlfnbm sc-hKgILt jdwhnD kCTcbZ"
      data-testid="ShapeFilter-rectangle"
      title="rectangle"
    >
      rectangle
    </li>
  </ul>
  <ul
    class="sc-gsTCUz nUtTt"
  >
    <li
      class="sc-dlfnbm sc-eCssSg jdwhnD gqWOnC"
      color="purple"
      data-testid="ColorFilter-purple"
      title="purple"
    >
      purple
    </li>
    <li
      class="sc-dlfnbm sc-eCssSg jdwhnD jruijj"
      color="red"
      data-testid="ColorFilter-red"
      title="red"
    >
      red
    </li>
    <li
      class="sc-dlfnbm sc-eCssSg jdwhnD cpXATr"
      color="green"
      data-testid="ColorFilter-green"
      title="green"
    >
      green
    </li>
    <li
      class="sc-dlfnbm sc-eCssSg jdwhnD eyGrwO"
      color="blue"
      data-testid="ColorFilter-blue"
      title="blue"
    >
      blue
    </li>
    <li
      class="sc-dlfnbm sc-eCssSg jdwhnD dNFxLw"
      color="yellow"
      data-testid="ColorFilter-yellow"
      title="yellow"
    >
      yellow
    </li>
    <li
      class="sc-dlfnbm sc-eCssSg jdwhnD lotLMF"
      color="grey"
      data-testid="ColorFilter-grey"
      title="grey"
    >
      grey
    </li>
  </ul>
</div>
`;
